Function-level optimization attributes.

eeckstein opened the pull request:
For now these are underscored attributes, i.e. compiler internal attributes:
    @_optimize(speed)
    @_optimize(size)
    @_optimize(none)
    
Those attributes override the command-line specified optimization mode for a specific function.
The @_optimize(none) attribute is equivalent to the already existing @_semantics("optimize.sil.never") attribute.

rdar://problem/35243425

eeckstein committed:
Unify the implementation of optimization mode in various option classes.

This commit is mostly refactoring.

*) Introduce a new OptimizationMode enum and use that in SILOptions and IRGenOptions
*) Allow the optimization mode also be specified for specific SILFunctions. This is not used in this commit yet and thus still a NFC.

Also, fixes a minor bug: we didn’t run mandatory IRGen passes for functions with @_semantics("optimize.sil.never")

eeckstein committed:
Function-level optimization attributes.

For now these are underscored attributes, i.e. compiler internal attributes:
@_optimize(speed)
@_optimize(size)
@_optimize(none)

Those attributes override the command-line specified optimization mode for a specific function.
The @_optimize(none) attribute is equivalent to the already existing @_semantics("optimize.sil.never") attribute

eeckstein committed:
IRGen: consider the function-level optimization mode for setting the minsize attribute on llvm functions.

eeckstein commented:
@swift-ci test

eeckstein commented:
@swift-ci smoke test

aschwaighofer reviewed (approved):
Looks good. Thank you.

shahmishal commented:
ci.swift.org is down for update, I will re-trigger this PR once update has been completed.

shahmishal commented:
@eeckstein  Due to merge conflict, I am not re-triggering the build.

slavapestov commented on include/swift/AST/Attr.def:
```diff
@@ -299,6 +299,10 @@ SIMPLE_DECL_ATTR(_implicitly_unwrapped_optional,
                  OnFunc | OnVar | OnSubscript | OnConstructor | RejectByParser, 72)
 
+DECL_ATTR(_optimize, Optimize,
+          OnFunc | OnConstructor | OnDestructor | OnSubscript |
```
Theoretically since you can apply this to subscripts, why not computed properties too?

eeckstein replied:
I'll add it. This is OnVar, right?

slavapestov commented:
@eeckstein This is a nice cleanup of the various flags, I like it. Can you replace existing usages of @_semantics("optimize.sil.never") in the test suite with @_optimize(none)? And why don't we get rid of the @_semantics variant altogether? It was never officially supported and from now on it's just unnecessary code to maintan.

aschwaighofer commented:
@slavapestov We use @_semantics for "array semantics" we cannot get rid of it until we have an alternative.

eeckstein commented:
@slavapestov Yes, let me do that in a follow-up commit

eeckstein committed:
Use the right decl for subscripts and computed properties to get the function attributes from.
Otherwise @_semantics, @_specialize  and @_optimize don't work for such declarations.

eeckstein commented:
Please test with following PR:
https://github.com/apple/swift-lldb/pull/256

@swift-ci smoke test

eeckstein commented:
rdar://problem/35243425
